pkg/datapath/linux/node.go

```diff
@@ -1184,6 +1184,8 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
 		oldIP4 = oldNode.GetNodeIP(false)
 		oldIP6 = oldNode.GetNodeIP(true)
 		oldKey = oldNode.EncryptionKey
+
+		n.deleteNodeIPs(oldNode.IPAddresses, newNode.IPAddresses)
```

Comment:
for deleteNodeIPs I was concerned about the case where a node updates its IP addresses and remove any no longer needed IPAddress. In existing code we would orphan the IP->nodeID mapping. So flow might look like this,

   1 NoadeA removes OldIP
   2 OldIP is no longer in NodeA->Addressses[]
   3 NodeA is deleted, unmap NodeA->Addresses[]
   4 OldIP from (1) is still in ip->nodeID map

Now that is at least a leak of nodeIDs. If,

   5. NodeB is added reusing OldIP NoceB->Addresses[... OldIP ...]
   6. allocateNodeID() finds the old nodeID
 
But events come out of order weird things might happen,

    5. PodAdd event before node update event with hostIP != OldIP we would create new nodeID and insert it
    6. NodeB is added we would scan Addresses[] and have two differet nodeIDs
 
once this happens something has to break because we can only have one xfrm nodeID mapping.

The text I wrote in the commit above is horrible by the way that should be rewritten. And also issue of how to actually trigger above hypothetical cases. I think adding/removing network interfaces might due it?

Reply:
I agree this is a bug to fix regardless of the impact. I'd just like to get to have some explanation for the XfrmOutPolBlock errors to be sure we have caught all observed bugs.

So let's look at what will happen if the changing node IP address is the NodeInternalIP. I'll detail the case with CiliumInternalIP tomorrow.

All is as seen from a node C, different from nodes A and B.

1. On the update to node A with change from OldIP to NewIP:
  a. We will reinstall XFRM configs for node A. This won't change anything because the NodeInternalIP isn't used in the XFRM configs.
  b. We will call `allocateIDForNode(nodeA)` as part of the node update. That will map NewIP to OldNodeID.
  c. We won't delete the old mapping, so OldIP will stay associated to OldNodeID as well.
2. On the update of pods on node A:
  a. We will update the ipcache entries for all pods of node A. While doing so, we will try to allocate a node ID for NewIP. That will return OldNodeID and ipcache entries for node A will thus have OldNodeID as before.
3. On the creation of node B with OldIP:
  a. We will call `allocateIDForNode(nodeB)` as part of the node creation. That will map all IPs of nodeB, including OldIP, to OldNodeID since the mapping OldIP <> OldNodeID still exists.
  b. We will install XFRM configs for node B. Those configs will use OldNodeID.
4. On the creation of pods on node B:
  a. We will create ipcache entries for all pods of node B. They will be using OldNodeID.

AFAICS, none of this will cause XfrmOutPolBlock.

The state after 4 will however cause us to encapsulate some encrypted packets with the wrong outer IP address (e.g., packets going to node B encapsulated with the IP of node A).

What could explain XfrmOutPolBlock is if we receive the update for pods on node A (2) before the update for node A (1). Then, we will allocate a new, different node ID for NewIP. Let's call it NewNodeID. When sending packets to pods of node A, we will set NewNodeID in the skb->mark, but the XFRM configs will still be matching on OldNodeID. Hence, XfrmOutPolBlock. It's the same bug I mentioned in [the comment below](https://github.com/cilium/cilium/pull/26725#discussion_r1260164732).